server/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -177,6 +187,38 @@ static boolean execute(SearchContext searchContext,
                 }
             }
 
+            // try to rewrite numeric or date sort to the optimized distanceFeatureQuery
+            if (searchContext.sort() != null) {
+                try {
+                    Query rewrittenQuery = tryRewriteNumericLongOrDateSort(searchContext, searcher.getIndexReader());
+                    if (rewrittenQuery != null) {
+                        query = new BooleanQuery.Builder()
+                            .add(query, BooleanClause.Occur.FILTER) // filter for original query
+                            .add(rewrittenQuery, BooleanClause.Occur.SHOULD) //should for rewrittenQuery
+                            .build();
+
+                        // modify sorts: add sort on _score as 1st sort, and move the sort on the original field as the last
+                        SortField[] oldSortFields = searchContext.sort().sort.getSort();
+                        DocValueFormat[] oldFormats = searchContext.sort().formats;
+                        SortField[] newSortFields = new SortField[oldSortFields.length + 1];
+                        DocValueFormat[] newFormats = new DocValueFormat[oldSortFields.length + 1];
+                        newSortFields[0] = SortField.FIELD_SCORE;
+                        newFormats[0] = DocValueFormat.RAW;
+                        for (int i = 1; i < oldSortFields.length; i++) {
+                            newSortFields[i] = oldSortFields[i];
+                            newFormats[i] = oldFormats[i];
+                        }
+                        // add the numeric field as the last sort, so we can access its values later
+                        newSortFields[oldSortFields.length] = oldSortFields[0];
+                        newFormats[oldSortFields.length] = oldFormats[0];
```

Comment:
Actually we should make it second, not last. In some cases, different values will be mapped to the same score because floats don't have infinite accuracy. So we should tie-break on the field that we are sorting on.

Reply:
@jpountz  thanks Adrien, corrected!